Fix on geometries and seisblock

ziyiyin97 opened the pull request:
Fix #85 

JUDI started to use Array{Array{Float32, 1}, 1} for geometry on yloc (a couple of versions ago) so the fix on rounding is necessary

mloubout reviewed (commented):
So gonna assume this particular case wasn't tested?

ziyiyin97 commented:
It is not tested because https://github.com/slimgroup/JUDI.jl/blob/16171876a379cc372b1d165483b1f1a3d790c984/src/TimeModeling/Types/judiVector.jl#L439 this line is wrong for the unit test shot records (which is 3D data)

ziyiyin97 commented:
I don't think there is any 2D shot record (i.e. yloc has different lengths than xloc or zloc) under data folder so the weird case is not tested. Let me know if we still want to add a test for it (then in this case, what would be a clean way to acquire this 2D shot record). I've added test to `src_to_SeisBlock`

mloubout commented:
You don't need a existing file, just create a block and check it's valid.

& commits for these changes is not good. Please rebase and keep the commit history clean

ziyiyin97 commented on src/TimeModeling/Types/GeometryStructure.jl:
```diff
@@ -179,12 +179,13 @@ function Geometry(data::SegyIO.SeisBlock; key="source", segy_depth_key="")
-            xloc[j] = convert(gt, xloc_full[traces][1])
-            yloc[j] = convert(gt,yloc_full[traces][1])
-            zloc[j] = abs.(convert(gt,zloc_full[traces][1]))
+            xloc[j] = convert(gt, xloc_full[traces][1:1])
```
It should contain `[1:1]` otherwise the `xloc_full[traces]` contains a repeated array, e.g. `[100.0, 100.0, 100.0]` but the `xloc[j]` of the source should really be `[100.0]`

mloubout replied:
That wouldn't be an issue with your change to `gt`

ziyiyin97 commented on test/test_geometry.jl:
```diff
@@ -47,9 +47,9 @@ datapath = joinpath(dirname(pathof(JUDI)))*"/../data/"
     @test isequal(typeof(src_geometry), GeometryIC{Float32})
     @test isequal(typeof(rec_geometry), GeometryIC{Float32})
-    @test isequal(get_header(block, "SourceSurfaceElevation")[1], src_geometry.zloc[1])
+    @test isequal(get_header(block, "SourceSurfaceElevation")[1], src_geometry.zloc[1][1])
```
Same above

mloubout replied:
Wuldn't be an issue without your changes

ziyiyin97 commented on test/test_judiVector.jl:
```diff
@@ -20,6 +20,11 @@ ftol = 1e-6
     info = example_info(nsrc=nsrc)
     dsize = (nsrc*nrec*ns, 1)
     rec_geometry = example_rec_geometry(nsrc=nsrc, nrec=nrec)
+    for i = 1:nsrc
```
I think the problem here is that there is a rounding in the locations due to working w/ SegyIO. If we just do a `==` test on `judiVector` then it will return false because of incompatible geometry

ziyiyin97 reviewed (commented):
Appreciate suggestion/critique on my comments

mloubout commented on src/TimeModeling/Types/GeometryStructure.jl:
```diff
@@ -161,23 +161,22 @@ end
     if key=="source"
         isempty(segy_depth_key) && (segy_depth_key="SourceSurfaceElevation")
         params = ["SourceX","SourceY",segy_depth_key]
-        gt = Float32
```
I still don't get why you need that and all the other changes seem to follow from it.

mloubout reviewed (commented):
I still don't get why some of these chabges are needed

ziyiyin97 commented:
See #85

mloubout commented on src/TimeModeling/Types/judiVector.jl:
```diff
@@ -432,7 +435,7 @@ function judiVector_to_SeisBlock(d::judiVector{avDT, AT}, q::judiVector{avDT, QT
         set_header!(blocks[j], "GroupX", convert(Array{Integer,1},round.(d.geometry.xloc[j]*1f3)))
         if length(d.geometry.yloc[j]) == 1
-            set_header!(blocks[j], "GroupY", Int(round(d.geometry.yloc[j]*1f3)))
+            set_header!(blocks[j], "GroupY", Int.(round.(d.geometry.yloc[j]*1f3)))
```
This is the only part relevant to #85, and it's not fixed correctly. Length is 1 so should just be `d.geometry.yloc[j][1]`. Doesn't explain the rest.

ziyiyin97 commented:
OK will fix that. But after fixing that, when you follow #85, you will find

```julia
julia> dobs.geometry.yloc
2-element Vector{Vector{Float32}}:
 [0.0]
 [0.0]
```

```julia
julia> judiVector(block).geometry.yloc
2-element Vector{Vector{Float32}}:
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
```

That's why we need to do this https://github.com/slimgroup/JUDI.jl/blob/0233edb25f605f56a90066ef35fce4b6622d701a/src/TimeModeling/Types/GeometryStructure.jl#L196

and then if we added this, we run the tests, we will get another error at https://github.com/slimgroup/JUDI.jl/blob/4675593bc4183a5a2ce34de73df75b4f946e6edb/test/test_judiVector.jl#L341

because 

```julia
julia> d_block.geometry.yloc
2-element Vector{Vector{Float32}}:
 [0.0]
 [0.0]
```

and

```julia
julia> d_get.geometry.yloc
2-element Vector{Vector{Float32}}:
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
 [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

```

That's why I fixed ``Geometry(::GeometryOOC)``. Then the next problem is at `src_geometry = Geometry(block; key="source", segy_depth_key="SourceSurfaceElevation")` https://github.com/slimgroup/JUDI.jl/blob/4675593bc4183a5a2ce34de73df75b4f946e6edb/test/test_judiVector.jl#L200 basically we get

```julia
julia> src_geometry.xloc
1-element Vector{Float32}:
 100.0

julia> src_geometry.yloc
1-element Vector{Float32}:
 0.0

julia> src_geometry.zloc
1-element Vector{Float32}:
 20.0
```

this is wrong because we expect to have `src_geometry.xloc` as `Vector{Vector{Float32}}`. That's why I edited the `Geometry(::SeisBlock)` and also edited the associated tests.

@mloubout basically I made these edits to make sure the `geometry.xloc` is `Vector{Vector{Float32}}, and `geometry.yloc` has only `[[0],[0]]` in 2D

mloubout commented:
Ok so everything about the coordinates makes sense thanks for the details.

I'll have another fresh look but overall I think that [1:1] and wavelet shape thingy is the only thing I think

ziyiyin97 commented:
Sure. I'll rebase after all things are sorted out

mloubout commented on src/TimeModeling/Types/judiVector.jl:
```diff
@@ -98,8 +97,12 @@ function judiVector(geometry::Geometry, data::Array{T, N}) where {T, N}
     return judiVector{T, Array{T, N}}("Seismic data vector", m, n, nsrc, geometry, dataCell)
 end
 
+function judiVector(geometry::Geometry, data::Vector{T}) where {T, N}
```
Don't see the point of this shape change.

ziyiyin97 replied:
This is to dispatch the data in judiVector so if 1D array is input, it reshapes it to 2D array. If you set up `q = judiVector(geometry, wavelet::Vector{Float32})` then

```julia
julia> # Nonlinear modeling
       dobs = Pr*F*adjoint(Ps)*q
┌ Warning: use values(kwargs) and keys(kwargs) instead of kwargs.data and kwargs.itr
│   caller = get_exp(::Int64, ::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}) at distributed.jl:52
└ @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:52
┌ Warning: `vendor()` is deprecated, use `BLAS.get_config()` and inspect the output instead
│   caller = npyinitialize() at numpy.jl:67
└ @ PyCall ~/.julia/packages/PyCall/L0fLP/src/numpy.jl:67
ERROR: TaskFailedException

    nested task error: TaskFailedException
    
        nested task error: MethodError: no method matching SincInterpolation(::Vector{Float32}, ::StepRangeLen{Float32, Float64, Float64, Int64}, ::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64})
        Closest candidates are:
          SincInterpolation(::Matrix{T}, ::Any, ::Any) where T<:Real at ~/.julia/dev/JUDI/src/TimeModeling/Utils/auxiliaryFunctions.jl:751
        Stacktrace:
         [1] time_resample(data::Vector{Float32}, geometry_in::GeometryIC{Float32}, dt_new::Float64)
           @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Utils/auxiliaryFunctions.jl:515
         [2] devito_interface(modelPy::PyCall.PyObject, model::Model, srcGeometry::GeometryIC{Float32}, srcData::Vector{Float32}, recGeometry::GeometryIC{Float32}, recData::Nothing, dm::Nothing, options::Options)
           @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/python_interface.jl:7
         [3] time_modeling(model_full::Model, srcGeometry::GeometryIC{Float32}, srcData::Vector{Float32}, recGeometry::GeometryIC{Float32}, recData::Nothing, dm::Nothing, op::Char, mode::Int64, options::Options)
           @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/time_modeling_serial.jl:37
         [4] (::Base.var"#76#78"{Base.var"#76#77#79"{ExponentialBackOff, Nothing, typeof(time_modeling)}})(::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
           @ Base ./error.jl:307
         [5] #76
           @ ./error.jl:290 [inlined]
         [6] #149
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:429 [inlined]
         [7] run_work_thunk(thunk::Distributed.var"#149#150"{Base.var"#76#78"{Base.var"#76#77#79"{ExponentialBackOff, Nothing, typeof(time_modeling)}}, Tuple{Model, GeometryIC{Float32}, Vector{Float32}, GeometryIC{Float32}, Nothing, Nothing, Char, Int64, Options}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, print_error::Bool)
           @ Distributed /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/process_messages.jl:63
         [8] run_work_thunk
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/process_messages.jl:72 [inlined]
         [9] (::Distributed.var"#96#98"{Distributed.RemoteValue, Distributed.var"#149#150"{Base.var"#76#78"{Base.var"#76#77#79"{ExponentialBackOff, Nothing, typeof(time_modeling)}}, Tuple{Model, GeometryIC{Float32}, Vector{Float32}, GeometryIC{Float32}, Nothing, Nothing, Char, Int64, Options}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}})()
           @ Distributed ./task.jl:423
        Stacktrace:
          [1] #remotecall_fetch#154
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:455 [inlined]
          [2] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
          [3] #remotecall_fetch#158
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
          [4] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
          [5] local_reduce!(my_future::Distributed.Future, other_future::Distributed.Future)
            @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:192
          [6] #149
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:429 [inlined]
          [7] run_work_thunk(thunk::Distributed.var"#149#150"{typeof(JUDI.local_reduce!), Tuple{Distributed.Future, Distributed.Future}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, print_error::Bool)
            @ Distributed /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/process_messages.jl:63
          [8] #remotecall_fetch#154
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
          [9] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
         [10] #remotecall_fetch#158
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [11] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [12] (::JUDI.var"#190#191"{Vector{Distributed.Future}, Int64})()
            @ JUDI ./task.jl:423
        Stacktrace:
         [1] #remotecall_fetch#154
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:455 [inlined]
         [2] remotecall_fetch
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
         [3] #remotecall_fetch#158
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [4] remotecall_fetch
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [5] (::JUDI.var"#190#191"{Vector{Distributed.Future}, Int64})()
           @ JUDI ./task.jl:423
    Stacktrace:
     [1] sync_end(c::Channel{Any})
       @ Base ./task.jl:381
     [2] macro expansion
       @ ./task.jl:400 [inlined]
     [3] reduce_level!(futures::Vector{Distributed.Future}, nleaf::Int64)
       @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:208
     [4] reduce!(futures::Vector{Distributed.Future})
       @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:242
     [5] (::JUDI.var"#187#189"{Vector{Any}, Int64, Vector{Distributed.Future}})()
       @ JUDI ./task.jl:423

...and 1 more exception.

Stacktrace:
  [1] sync_end(c::Channel{Any})
    @ Base ./task.jl:381
  [2] macro expansion
    @ ./task.jl:400 [inlined]
  [3] task_distributed(::Function, ::Distributed.WorkerPool, ::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:132
  [4] task_distributed(::Function, ::Distributed.WorkerPool, ::Model, ::Vararg{Any})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:125
  [5] time_modeling(::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:4
  [6] time_modeling
    @ ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:4 [inlined]
  [7] (::JUDI.var"#221#223"{Options, Info, Model, GeometryIC{Float32}, GeometryIC{Float32}, UnitRange{Int64}})(src::judiVector{Float32, Vector{Float32}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/LinearOperators/judiPDEfull.jl:52
  [8] *(A::judiPDEfull{Float32, Float32}, v::judiVector{Float32, Vector{Float32}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/LinearOperators/judiPDEfull.jl:104
  [9] afoldl
    @ ./operators.jl:612 [inlined]
 [10] *(a::judiProjection{Float32, Float32}, b::judiModeling{Float32, Float32}, c::judiProjection{Float32, Float32}, xs::judiVector{Float32, Vector{Float32}})
    @ Base ./operators.jl:655
 [11] top-level scope
    @ REPL[40]:2
```

mloubout replied:
So instead of just fixing `SincInterpolation` to work on vector you enforce painful reshape everywhere?

ziyiyin97 replied:
Let me put it this way, if `wavelet::Vector{T}`:

`q1 = judiVector(geometry, wavelet)`

and

`q2 = judiVector(geometry, reshape(wavelet, :, 1))`

should be the same source function. The problem is now they are NOT. So this reshape should be enforced so that `q.data` is always a `Matrix` instead of a `Vector`.

mloubout replied:
No! This reshape is enforced "by you" and the force reshape you introduce then you change the test to pretend that change isn't there and messing things up.
Both `q1` and `q2` are valid and no it should not be the same since it's a different input, there is zero reason to enforce they the same.

ziyiyin97 replied:
I would argue one more time on this:

if we want to save q1 and q2 as SEGY files, we have to dispatch there again (because `q1.data::Vector` and `q2.data::Matrix`). Now this question comes again: should `segy(q1)` be the same as `segy(q2)`? There might be a lot of problems associated with this so I find it quite weird to make things like this, i.e. source functions not having a standardized form.

We might experience other issues later if we further develop some new features. We will always face the problem that we are uncertain about the type of `q.data`. Why don't we enforce it to be a `Matrix` which can really save a lot of dispatches in the future, and things could be in standard forms?

This is my personal opinion but feel free to convince me if this isn't a good practice.

mloubout replied:
You are enforcing implemetation details onto the user which is wrong. The dispatch is *made* for different types to to prevent it, that's why it's called multiple dispatch, mot "single dispatch or we error". 

`Vector` and `Matrix` are `::Array{T, N}` so your dispatch is fixed, done. This whole PR is literally two line change
- FIxe Sinc to accept Vector (which is what dispatch is for)
- Add the `[1]` for sizxe one
And that's it. Everything else is convoluted way to bypass a "single type enforcing" rather than a "diaptch".

ziyiyin97 replied:
OK then the problem is that `judiVector->segy->judiVector` isn't identity operation. We don't want that

ziyiyin97 replied:
> But enforcing SEGY limitation into judiVector is backward.

I would argue this also entails the forward process -- if `judiVector(geometry, data::Vector)` and `judiVector(geometry, data::Matrix)` end up to have the same SEGY file, then you have no way to ensure backward compatibility

mloubout replied:
That's not what backward compatibility means.

I only meant backward that it's the wrong way to design it.

ziyiyin97 replied:
OK then we should do:

We accept `data::Vector` for `judiVector` but when we load SegyIO file, we always make `data::Matrix` for `judiVector`. Does that sound good?

mloubout replied:
Why? The current version reads it as a Vector what's the issue with it.

mloubout commented on test/test_linear_operators.jl:
```diff
@@ -135,7 +135,7 @@ end
     @test isequal(J.recGeometry, rec_geometry)
     @test isequal(J.source.geometry, src_geometry)
-    @test all(isequal(J.source.data[i], wavelet) for i=1:nsrc)
+    @test all(isequal(J.source.data[i], reshape(wavelet, src_geometry.nt[1], 1)) for i=1:nsrc)
```
wouldn't be needed without that reshape.

ziyiyin97 replied:
This is the `reshape` that corresponds to the case above

mloubout reviewed (commented):
You haven't change anything since last review.

ziyiyin97 commented:
`srcGeometry.xloc` should be `Vector{Vector{T}}`, e.g. `[[100f0],[200f0]]` but now it's not -- https://github.com/slimgroup/JUDI.jl/pull/69#discussion_r791809809 you can see here that now it is `Vector{T}`. We should enforce the source and receive geometries to be set in a standard way so that converting it to SEGY file and back things are still reasonable. We don't want 2 judiVectors with different geometries while they actually correspond to the same source, right?

mloubout commented:
super-seeded by #91
